### PR TITLE
Avoid stackoverflow for deeply nested objects and arrays.

### LINKED ISSUE #18: Stack overflow error caused by openjson parsing of untrusted JSON String
# Stack overflow error caused by openjson parsing of untrusted JSON String

## Description

Using **openjson** to parse untrusted JSON String may be vulnerable to denial of service (DOS) attacks. If the parser is running on user supplied input, an attacker may supply content that causes the parser to crash by stackoverflow.

## Error Log

```
Exception in thread "main" java.lang.StackOverflowError
	at com.github.openjson.JSONArray.<init>(JSONArray.java:54)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:426)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)
	at com.github.openjson.JSONTokener.nextValue(JSONTokener.java:120)
	at com.github.openjson.JSONTokener.readArray(JSONTokener.java:450)

```

## PoC
```xml
        <dependency>
            <groupId>com.github.openjson</groupId>
            <artifactId>openjson</artifactId>
            <version>1.0.11</version>
        </dependency>
```

```java
import com.github.openjson.JSONArray;
import com.github.openjson.JSONObject;
import com.github.openjson.JSONTokener;

/**
 * PoC
 *
 * @since 1.0.0
 */
public class PoC {
    public final static int TOO_DEEP_NESTING = 9999;
    public final static String TOO_DEEP_Array = _nestedDoc(TOO_DEEP_NESTING, "[ ", "] ", "1");
    public final static String TOO_DEEP_Object = "{" + _nestedDoc(TOO_DEEP_NESTING, "\"a\": { ", "} ", "1") + "}";

    public static String _nestedDoc(int nesting, String open, String close, String content) {
        StringBuilder sb = new StringBuilder(nesting * (open.length() + close.length()));
        for (int i = 0; i < nesting; ++i) {
            sb.append(open);
            if ((i & 31) == 0) {
                sb.append("\n");
            }
        }
        sb.append("\n").append(content).append("\n");
        for (int i = 0; i < nesting; ++i) {
            sb.append(close);
            if ((i & 31) == 0) {
                sb.append("\n");
            }
        }
        return sb.toString();
    }
    public static void main(String[] args) {
        JSONTokener jsonTokener = new JSONTokener(TOO_DEEP_Array);
        JSONArray jsonArray = new JSONArray(jsonTokener);
        JSONTokener jsonTokener2 = new JSONTokener(TOO_DEEP_Object);
        JSONObject jsonObject = new JSONObject(jsonTokener2);
    }
}
```

## Rectification Solution
1. Refer to the solution of jackson-databind: Add the depth variable to record the current parsing depth. If the parsing depth exceeds a certain threshold, an exception is thrown. (https://github.com/FasterXML/jackson-databind/commit/fcfc4998ec23f0b1f7f8a9521c2b317b6c25892b)

2. Refer to the GSON solution: Change the recursive processing on deeply nested arrays or JSON objects to stack+iteration processing.(（https://github.com/google/gson/commit/2d01d6a20f39881c692977564c1ea591d9f39027）)

### PR DESCRIPTION
**Pull request by mfursov**
Fixes https://github.com/openjson/openjson/issues/18

Guys if nobody minds I can merge this fix & create a new version in Maven.

### CODE REVIEW

**Review by solomax — APPROVED** [review]
Shall the original PoC be added as unit test?

**Comment by mfursov** [comment]
> Shall the original PoC be added as unit test?

I do not mind if you or the the author add it into this or a new PR.

**Comment by solomax** [comment]
I'm currerly on vacation without laptop :( only can read and review ;)

File: src/main/java/com/github/openjson/JSONTokener.java [hunk 1]
```diff
@@ -56,14 +56,22 @@
...
  */
 public class JSONTokener {
 
+    /**
+     * Maximum allowed JSON nesting level supported by the parser.
+     * Used to avoid StackOverflow errors.
+     * For backward compatibility, the value can be changed (globally) using a system property.
+     */
+    public static final int MAX_NESTING_LEVEL =
+            Integer.parseInt(System.getProperty("com.github.openjson.MAX_NESTING_LEVEL", "100"));
```
Comment (martin-g) [inline src/main/java/com/github/openjson/JSONTokener.java:73]:
```suggestion
            Integer.getInteger("com.github.openjson.MAX_NESTING_LEVEL", 100);
```

File: src/main/java/com/github/openjson/JSONTokener.java [hunk 2]
```diff
@@ -108,16 +116,24 @@ public JSONTokener(Reader input) throws IOException {
      * @throws JSONException if the input is malformed.
      */
     public Object nextValue() throws JSONException {
+        return nextValue(0);
+    }
+
+    private Object nextValue(int nestingLevel) throws JSONException {
+        if (nestingLevel >= MAX_NESTING_LEVEL) {
+            throw new JSONException("Failed to parse input. JSONTokener. " +
```
Comment (martin-g) [inline src/main/java/com/github/openjson/JSONTokener.java:124]:
```suggestion
            throw new JSONException("Failed to parse input. JSONTokener." +
```

File: src/test/java/com/github/openjson/JSONTokenerTest.java [hunk 3]
```diff
@@ -616,4 +616,63 @@ public void testDehexchar() {
...
+        }
+        assertNotNull(throwable);
+        assertTrue(throwable instanceof JSONException);
+        assertTrue(throwable.getMessage().contains("MAX_NESTING_LEVEL"));
+    }
+
+    private static String generateNestedArray(int depth) {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < depth; i++) {
+            sb.append("[");
```
Comment (martin-g) [inline src/test/java/com/github/openjson/JSONTokenerTest.java:641]:
```suggestion
            sb.append('[');
```

File: src/test/java/com/github/openjson/JSONTokenerTest.java [hunk 4]
```diff
@@ -616,4 +616,63 @@ public void testDehexchar() {
...
+        }
+        for (int i = 0; i < depth; i++) {
+            sb.append("]");
```
Comment (martin-g) [inline src/test/java/com/github/openjson/JSONTokenerTest.java:644]:
```suggestion
            sb.append(']');
```

File: src/test/java/com/github/openjson/JSONTokenerTest.java [hunk 5]
```diff
@@ -616,4 +616,63 @@ public void testDehexchar() {
...
+        assertTrue(throwable.getMessage().contains("MAX_NESTING_LEVEL"));
+    }
+
+    private static String generateNestedObject(int depth) {
+        StringBuilder sb = new StringBuilder("{");
+        for (int i = 0; i < depth - 1; i++) {
+            sb.append("\"a\":{");
+        }
+        for (int i = 0; i < depth + 1; i++) {
+            sb.append("}");
```
Comment (martin-g) [inline src/test/java/com/github/openjson/JSONTokenerTest.java:673]:
```suggestion
            sb.append('}');
```

File: src/test/java/com/github/openjson/JSONTokenerTest.java [hunk 6]
```diff
@@ -616,4 +616,63 @@ public void testDehexchar() {
...
+        }
+        sb.append("}");
```
Comment (martin-g) [inline src/test/java/com/github/openjson/JSONTokenerTest.java:675]:
```suggestion
        sb.append('}');
```

**Review by martin-g — APPROVED** [review]

**Comment by mfursov** [comment]
https://repo1.maven.org/maven2/com/github/openjson/openjson/1.0.13/

**Pull request merged**